python/ray/tune/execution/trial_runner.py

```diff
@@ -141,11 +148,32 @@ def __init__(
         self._remote_dir = remote_dir
 
         self._last_checkpoint_time = 0.0
+        self._last_sync_time = 0.0
+        self._trial_checkpoint_config = trial_checkpoint_config
+        self._trial_num_checkpoints_since_last_sync: DefaultDict[
+            Trial, int
+        ] = defaultdict(int)
+        self._excessive_sync_threshold = float(
+            os.environ.get(
+                "TUNE_WARN_EXCESSIVE_EXPERIMENT_CHECKPOINT_SYNC_THRESHOLD_S", "30"
+            )
+        )
 
     @property
     def auto_checkpoint_enabled(self):
         return self._auto_checkpoint_enabled
 
+    def on_trial_checkpoint(self, trial: Trial):
+        if not self._trial_checkpoint_config.num_to_keep:
+            return
+        self._trial_num_checkpoints_since_last_sync[trial] += 1
+
+    def should_force_cloud_sync(self) -> bool:
+        return any(
+            num_checkpoints > self._trial_checkpoint_config.num_to_keep
+            for num_checkpoints in self._trial_num_checkpoints_since_last_sync.values()
+        )
```

Comment:
This could be costly if we have a lot of trials running.

Can we move the force check to `on_trial_checkpoint`? We can just set an attribute `self._force_next_sync` if we exceed the threshold. Then we avoid a O(n) scan through all trials everytime we call `self.checkpoint()` (which is every step)